### ISSUE TITLE
(How) should unreachable switch cases be reported?

### ISSUE THREAD

**Issue by johnniwinther** [dart-lang/language#2924]
The full exhaustiveness checking algorithm has the ability to find switch cases that are unreachable because they are covered by previous cases. (How) should we report these?

Currently the analyzer reports this as hints, which can be turned off (ignored). The CFE doesn't have an equivalent and can only report these as errors or warnings, neither of which can be turned off individually.

@munificent @leafpetersen @lrhn @stereotype441 @scheglov

**Comment by lrhn** [comment]
My opinion: Report no warning that cannot be silenced.
Which currently means report no error outside of the analyzer.

We don't generally prevent (make an error) dead code, so we probably shouldn't here either.

**Comment by jakemac53** [comment]
I don't see any difference between this and the general dead code, so I would just treat it exactly the same as other dead code (I think those are compiler warnings, but not sure).

**Comment by munificent** [comment]
I agree. I think "warning" as a category should always be used for what is essentially dead code, and an unreachable case is dead code. I'd make it a warning.

**Comment by lrhn** [comment]
Dead code is mostly "info":
```dart
void main() {
  String s = "here";
  print(s ?? "unreachable");
  if (true) {
    print("here");
  } else {
    print("unreachable");
  }
  return;
  print("unreachable");
}
``` 
in DartPad shows:
```
warning
line 3 • The left operand can't be null, so the right operand is never executed.[ (view docs)](https://dart.dev/diagnostics/dead_null_aware_expression)
Try removing the operator and the right operand.
info
line 6 • Dead code.[ (view docs)](https://dart.dev/diagnostics/dead_code)
Try removing the code, or fixing the code before it so that it can be reached.
info
line 10 • Dead code.[ (view docs)](https://dart.dev/diagnostics/dead_code)
Try removing the code, or fixing the code before it so that it can be reached.
```

And this only shows in the analyzer, nothing from compilers.

**Comment by johnniwinther** [comment]
I think the risky part here about the terminology is that if we equate "warnings" with some that is safe to report because it doesn't stop compilation, then the CFE should _not_ emit warnings, because they do stop compilation in some scenarios (building the sdk, for instance).

**Comment by lrhn** [comment]
Emitting warnings should be safe. The entire point of warnings is that they are *not* errors.

Turning warnings into build-breaking errors is a choice that some projects can take, but it should always be entirely on them.
Coming back and saying "Your warning broke my build" should be met with a "Well, you asked for it!".

That said, that's not what people are actually doing. Until then, I don't think the CFE should emit any warnings that cannot be ignored. And since the CFE doesn't honor `// ignore: ...`, that means not emitting any warnings.

(An alternative would be a global flag controlling whether any warnings are emitted. The default can be on or off. That still won't stop someone from turning it on and breaking their own build if they see a warning.)
